File: UMeshCore/src/Mesh/MeshValidator.cpp
#include "umeshcore/Mesh/MeshValidator.h"

#include <algorithm>
#include <cmath>
#include <cstdio>
#include <limits>
#include <unordered_map>
#include <unordered_set>

#include "umeshcore/Mesh/MeshKernel.h"
#include "umeshcore/Mesh/MeshPredicates.h"
#include "umeshcore/Mesh/MeshTypes.h"

namespace umeshcore::MeshValidator {

std::string invariantCode(Invariant i) {
    switch (i) {
        case Invariant::IndexBounds: return "I1";
        case Invariant::Degenerate: return "I2";
        case Invariant::Winding: return "I3";
        case Invariant::Manifold: return "I4";
        case Invariant::BoundaryMatchesRings: return "I5";
        case Invariant::Coverage: return "I7";
        case Invariant::OrphanVertices: return "I8";
        case Invariant::Constraints: return "I9";
        case Invariant::TrianglesInHoles: return "I10";
        case Invariant::IndexCeiling: return "I11";
    }
    return "?";
}

std::string Report::summary() const {
    if (isValid() && warnings.empty()) return "valid";
    std::string out;
    bool first = true;
    for (const auto& [inv, msg] : failures) {
        if (!first) out += "; ";
        out += invariantCode(inv) + ": " + msg;
        first = false;
    }
    for (const auto& [inv, msg] : warnings) {
        if (!first) out += "; ";
        out += invariantCode(inv) + " (warning): " + msg;
        first = false;
    }
    return out;
}

std::string Report::shortStatus() const {
    if (isValid()) {
        if (warnings.empty()) return "Watertight";
        return "Watertight, " + std::to_string(warnings.size()) + " note(s)";
    }
    return std::to_string(failures.size()) + " problem(s)";
}

namespace {
bool edgeIsBefore(const MeshEdge& lhs, const MeshEdge& rhs) {
    return lhs.a != rhs.a ? lhs.a < rhs.a : lhs.b < rhs.b;
}
} // namespace

Report validate(
    const std::vector<Vec2>& points, const std::vector<std::uint16_t>& triangles,
    const std::vector<int>& outer, const std::vector<std::vector<int>>& holes,
    const std::vector<std::pair<int, int>>& constraints, double coverageTolerance) {
    Report report;
    const std::size_t n = points.size();

    // I1 -- index bounds and triangle count.
    if (triangles.size() % 3 != 0) {
        report.fail(
            Invariant::IndexBounds,
            "triangle list length " + std::to_string(triangles.size()) + " is not a multiple of 3");
        return report;
    }
    for (std::uint16_t index : triangles) {
        if (static_cast<std::size_t>(index) >= n) {
            report.fail(
                Invariant::IndexBounds,
                "index " + std::to_string(index) + " out of range (0.." + std::to_string(n - 1) + ")");
            return report;
        }
    }

    std::vector<std::tuple<int, int, int>> tris;
    tris.reserve(triangles.size() / 3);
    for (std::size_t t = 0; t < triangles.size(); t += 3) {
        tris.emplace_back(triangles[t], triangles[t + 1], triangles[t + 2]);
    }

    // I2 -- no degenerate triangles, and no repeated vertex within one.
    for (const auto& [a, b, c] : tris) {
        if (a == b || b == c || a == c) {
            report.fail(
                Invariant::Degenerate,
                "triangle (" + std::to_string(a) + "," + std::to_string(b) + "," + std::to_string(c) +
                    ") repeats a vertex");
        } else if (MeshPredicates::orient2d(points[a], points[b], points[c]) == 0.0) {
            report.fail(
                Invariant::Degenerate,
                "triangle (" + std::to_string(a) + "," + std::to_string(b) + "," + std::to_string(c) +
                    ") has zero area");
        }
    }

    // I3 -- consistent winding.
    {
        std::unordered_set<int> signs;
        for (const auto& [a, b, c] : tris) {
            const double o = MeshPredicates::orient2d(points[a], points[b], points[c]);
            if (o != 0.0) signs.insert(o > 0.0 ? 1 : -1);
        }
        if (signs.size() > 1) {
            report.fail(Invariant::Winding, "triangles do not share a consistent winding");
        }
    }

    // I4 -- manifold: every undirected edge used once (boundary) or twice (interior).
    std::unordered_map<MeshEdge, int, MeshEdgeHash> edgeCount;
    edgeCount.reserve(tris.size() * 3);
    for (const auto& [a, b, c] : tris) {
        const std::pair<int, int> edges[3] = {{a, b}, {b, c}, {c, a}};
        for (const auto& [u, v] : edges) {
            edgeCount[MeshEdge(static_cast<std::uint16_t>(u), static_cast<std::uint16_t>(v))] += 1;
        }
    }
    {
        std::vector<MeshEdge> overused;
        for (const auto& [edge, count] : edgeCount) {
            if (count > 2) overused.push_back(edge);
        }
        if (!overused.empty()) {
            const MeshEdge first = *std::min_element(overused.begin(), overused.end(), edgeIsBefore);
            report.fail(
                Invariant::Manifold,
                std::to_string(overused.size()) + " edge(s) shared by more than 2 triangles, e.g. (" +
                    std::to_string(first.a) + "," + std::to_string(first.b) + ")");
        }
    }

    // I5 -- the boundary of the triangulation is exactly the ring set.
    std::unordered_set<MeshEdge, MeshEdgeHash> boundary;
    for (const auto& [edge, count] : edgeCount) {
        if (count == 1) boundary.insert(edge);
    }
    std::unordered_set<MeshEdge, MeshEdgeHash> expected;
    {
        std::vector<const std::vector<int>*> allRings;
        allRings.push_back(&outer);
        for (const auto& h : holes) allRings.push_back(&h);
        for (const auto* ringPtr : allRings) {
            const auto& ring = *ringPtr;
            if (ring.size() < 2) continue;
            for (std::size_t i = 0; i < ring.size(); ++i) {
                expected.insert(MeshEdge(
                    static_cast<std::uint16_t>(ring[i]),
                    static_cast<std::uint16_t>(ring[(i + 1) % ring.size()])));
            }
        }
    }
    {
        std::vector<MeshEdge> missing;
        for (const auto& e : expected) {
            if (!boundary.contains(e)) missing.push_back(e);
        }
        std::vector<MeshEdge> extra;
        for (const auto& e : boundary) {
            if (!expected.contains(e)) extra.push_back(e);
        }
        if (!missing.empty()) {
            const MeshEdge first = *std::min_element(missing.begin(), missing.end(), edgeIsBefore);
            report.fail(
                Invariant::BoundaryMatchesRings,
                std::to_string(missing.size()) + " outline edge(s) are not on the mesh boundary, e.g. (" +
                    std::to_string(first.a) + "," + std::to_string(first.b) +
                    ") -- this is a hole or a torn edge");
        }
        if (!extra.empty()) {
            const MeshEdge first = *std::min_element(extra.begin(), extra.end(), edgeIsBefore);
            report.fail(
                Invariant::BoundaryMatchesRings,
                std::to_string(extra.size()) + " boundary edge(s) are not outline edges, e.g. (" +
                    std::to_string(first.a) + "," + std::to_string(first.b) +
                    ") -- this is a gap inside the region");
        }
    }

    // I7 -- coverage. The invariant that would have caught the shipped bug.
    double covered = 0.0;
    for (const auto& [a, b, c] : tris) {
        covered += std::abs(MeshPredicates::signedArea2(points[a], points[b], points[c])) / 2.0;
    }
    double target = MeshKernel::polygonArea(points, outer);
    for (const auto& hole : holes) target -= MeshKernel::polygonArea(points, hole);
    if (target > 0.0) {
        const double relative = std::abs(covered - target) / target;
        if (relative > coverageTolerance) {
            const double percent = 100.0 * covered / target;
            char buf[128];
            std::snprintf(buf, sizeof(buf), "coverage %.4f %% of the region (%.6f vs %.6f)", percent, covered, target);
            report.fail(Invariant::Coverage, buf);
        }
    }

    // I8 -- no orphan outline vertices.
    std::unordered_set<int> referenced;
    for (std::uint16_t idx : triangles) referenced.insert(idx);
    std::unordered_set<int> ringVertices(outer.begin(), outer.end());
    for (const auto& hole : holes) ringVertices.insert(hole.begin(), hole.end());
    {
        std::vector<int> orphans;
        for (int v : ringVertices) {
            if (!referenced.contains(v)) orphans.push_back(v);
        }
        if (!orphans.empty()) {
            const int first = *std::min_element(orphans.begin(), orphans.end());
            report.fail(
                Invariant::OrphanVertices,
                std::to_string(orphans.size()) +
                    " outline vertex/vertices missing from the triangulation, e.g. " +
                    std::to_string(first));
        }
    }

    // I9 -- declared constraints are present as edges.
    for (const auto& [a, b] : constraints) {
        const MeshEdge key(static_cast<std::uint16_t>(a), static_cast<std::uint16_t>(b));
        if (!edgeCount.contains(key)) {
            report.fail(
                Invariant::Constraints,
                "constraint edge (" + std::to_string(key.a) + "," + std::to_string(key.b) +
                    ") is absent from the triangulation");
        }
    }

    // I10 -- no triangle centroid inside a hole.
    for (const auto& hole : holes) {
        for (const auto& [a, b, c] : tris) {
            const Vec2 centroid = (points[a] + points[b] + points[c]) / 3.0f;
            if (MeshKernel::pointInRing(points, hole, centroid)) {
                report.fail(
                    Invariant::TrianglesInHoles,
                    "triangle (" + std::to_string(a) + "," + std::to_string(b) + "," + std::to_string(c) +
                        ") lies inside a hole");
                break;
            }
        }
    }

    // I11 -- the uint16 index ceiling the file format imposes.
    if (n > static_cast<std::size_t>(std::numeric_limits<std::uint16_t>::max())) {
        report.fail(
            Invariant::IndexCeiling,
            std::to_string(n) + " vertices exceeds the uint16 index limit of 65535");
    }

    return report;
}

double coverageRatio(
    const std::vector<Vec2>& points, const std::vector<std::uint16_t>& triangles,
    const std::vector<int>& outer, const std::vector<std::vector<int>>& holes) {
    double covered = 0.0;
    for (std::size_t t = 0; t + 2 < triangles.size(); t += 3) {
        const int a = triangles[t], b = triangles[t + 1], c = triangles[t + 2];
        if (static_cast<std::size_t>(a) >= points.size() || static_cast<std::size_t>(b) >= points.size() ||
            static_cast<std::size_t>(c) >= points.size()) {
            continue;
        }
        covered += std::abs(MeshPredicates::signedArea2(points[a], points[b], points[c])) / 2.0;
    }
    double target = MeshKernel::polygonArea(points, outer);
    for (const auto& hole : holes) target -= MeshKernel::polygonArea(points, hole);
    return target > 0.0 ? covered / target : 0.0;
}

} // namespace umeshcore::MeshValidator

File: UltraMesh 2d animation/Data/MeshValidator.swift
import Foundation
import simd

/// Invariants a mesh must satisfy, checked after every mutation.
///
/// The point is that a hole becomes impossible to *ship*, not merely unlikely to
/// be drawn. Before this type existed there was no coverage check anywhere in
/// the editor, which is why a triangle list covering 97.8 % of a silhouette went
/// straight through to the renderer, the project file and the exporters without
/// a single complaint. `sanitizedForRender` could not catch it either: its
/// rebuild only fires when `indices` is *completely* empty, so a
/// partially-covered list passes through untouched.
///
/// `.coverage` is the invariant that catches that specific defect. The others
/// catch the failure modes a triangulation can exhibit while still having the
/// right total area — a torn edge, a flipped triangle, a vertex that quietly
/// stopped being referenced.
///
/// Normative reference: `Meshing/pymesh/validate.py`.
enum MeshValidator {

    /// Invariant identifiers, kept stable so they can be quoted in a bug report.
    enum Invariant: String, Equatable {
        case indexBounds       = "I1"
        case degenerate        = "I2"
        case winding           = "I3"
        case manifold          = "I4"
        case boundaryMatchesRings = "I5"
        case coverage          = "I7"
        case orphanVertices    = "I8"
        case constraints       = "I9"
        case trianglesInHoles  = "I10"
        case indexCeiling      = "I11"
    }

    struct Report {
        private(set) var failures: [(Invariant, String)] = []
        private(set) var warnings: [(Invariant, String)] = []

        var isValid: Bool { failures.isEmpty }

        mutating func fail(_ invariant: Invariant, _ message: String) {
            failures.append((invariant, message))
        }

        mutating func warn(_ invariant: Invariant, _ message: String) {
            warnings.append((invariant, message))
        }

        /// One-line summary suitable for a log or an inspector tooltip.
        var summary: String {
            if isValid && warnings.isEmpty { return "valid" }
            var parts = failures.map { "\($0.0.rawValue): \($0.1)" }
            parts += warnings.map { "\($0.0.rawValue) (warning): \($0.1)" }
            return parts.joined(separator: "; ")
        }

        /// Short phrase for the health indicator next to the vertex count.
        var shortStatus: String {
            if isValid { return warnings.isEmpty ? "Watertight" : "Watertight, \(warnings.count) note(s)" }
            return "\(failures.count) problem(s)"
        }
    }

    /// Check every invariant against a triangle list.
    ///
    /// - Parameter coverageTolerance: relative area tolerance. The default is
    ///   tight on purpose: the pipeline is exact by construction, so anything
    ///   above rounding noise means real geometry went missing.
    static func validate(
        points: [SIMD2<Float>],
        triangles: [UInt16],
        outer: [Int],
        holes: [[Int]] = [],
        constraints: [(Int, Int)] = [],
        coverageTolerance: Double = 1e-6
    ) -> Report {
        var report = Report()
        let n = points.count

        // I1 — index bounds and triangle count.
        guard triangles.count % 3 == 0 else {
            report.fail(.indexBounds,
                        "triangle list length \(triangles.count) is not a multiple of 3")
            return report
        }
        for index in triangles where Int(index) >= n {
            report.fail(.indexBounds, "index \(index) out of range (0..\(n - 1))")
            return report
        }

        var tris: [(Int, Int, Int)] = []
        tris.reserveCapacity(triangles.count / 3)
        for t in stride(from: 0, to: triangles.count, by: 3) {
            tris.append((Int(triangles[t]), Int(triangles[t + 1]), Int(triangles[t + 2])))
        }

        // I2 — no degenerate triangles, and no repeated vertex within one.
        for (a, b, c) in tris {
            if Set([a, b, c]).count != 3 {
                report.fail(.degenerate, "triangle (\(a),\(b),\(c)) repeats a vertex")
            } else if MeshPredicates.orient2d(points[a], points[b], points[c]) == 0 {
                report.fail(.degenerate, "triangle (\(a),\(b),\(c)) has zero area")
            }
        }

        // I3 — consistent winding.
        var signs = Set<Int>()
        for (a, b, c) in tris {
            let o = MeshPredicates.orient2d(points[a], points[b], points[c])
            if o != 0 { signs.insert(o > 0 ? 1 : -1) }
        }
        if signs.count > 1 {
            report.fail(.winding, "triangles do not share a consistent winding")
        }

        // I4 — manifold: every undirected edge used once (boundary) or twice (interior).
        var edgeCount: [MeshEdge: Int] = [:]
        edgeCount.reserveCapacity(tris.count * 3)
        for (a, b, c) in tris {
            for (u, v) in [(a, b), (b, c), (c, a)] {
                edgeCount[MeshEdge(UInt16(u), UInt16(v)), default: 0] += 1
            }
        }
        let overused = edgeCount.filter { $0.value > 2 }
        if let first = overused.keys.sorted(by: edgeIsBefore).first {
            report.fail(.manifold,
                        "\(overused.count) edge(s) shared by more than 2 triangles, "
                        + "e.g. (\(first.a),\(first.b))")
        }

        // I5 — the boundary of the triangulation is exactly the ring set.
        let boundary = Set(edgeCount.filter { $0.value == 1 }.keys)
        var expected = Set<MeshEdge>()
        for ring in [outer] + holes where ring.count >= 2 {
            for i in ring.indices {
                expected.insert(MeshEdge(UInt16(ring[i]), UInt16(ring[(i + 1) % ring.count])))
            }
        }
        let missing = expected.subtracting(boundary)
        let extra = boundary.subtracting(expected)
        if let first = missing.sorted(by: edgeIsBefore).first {
            report.fail(.boundaryMatchesRings,
                        "\(missing.count) outline edge(s) are not on the mesh boundary, "
                        + "e.g. (\(first.a),\(first.b)) — this is a hole or a torn edge")
        }
        if let first = extra.sorted(by: edgeIsBefore).first {
            report.fail(.boundaryMatchesRings,
                        "\(extra.count) boundary edge(s) are not outline edges, "
                        + "e.g. (\(first.a),\(first.b)) — this is a gap inside the region")
        }

        // I7 — coverage. The invariant that would have caught the shipped bug.
        // Uses signedArea2, not orient2d: orient2d answers a yes/no question and
        // is free to report +/-1 rather than the determinant, so summing areas
        // with it would count a near-degenerate triangle as half a unit of area.
        var covered = 0.0
        for (a, b, c) in tris {
            covered += abs(MeshPredicates.signedArea2(points[a], points[b], points[c])) / 2.0
        }
        var target = MeshKernel.polygonArea(points: points, ring: outer)
        for hole in holes { target -= MeshKernel.polygonArea(points: points, ring: hole) }
        if target > 0 {
            let relative = abs(covered - target) / target
            if relative > coverageTolerance {
                let percent = 100.0 * covered / target
                report.fail(.coverage,
                            String(format: "coverage %.4f %% of the region (%.6f vs %.6f)",
                                   percent, covered, target))
            }
        }

        // I8 — no orphan outline vertices. Coverage cannot see this: a vertex
        // collinear with its neighbours contributes no area, so it can vanish
        // while the total still reads 100 %, taking its bone weights and every
        // deform keyframe that addressed it.
        let referenced = Set(triangles.map { Int($0) })
        var ringVertices = Set(outer)
        for hole in holes { ringVertices.formUnion(hole) }
        let orphans = ringVertices.subtracting(referenced)
        if let first = orphans.sorted().first {
            report.fail(.orphanVertices,
                        "\(orphans.count) outline vertex/vertices missing from the "
                        + "triangulation, e.g. \(first)")
        }

        // I9 — declared constraints are present as edges.
        for (a, b) in constraints {
            let key = MeshEdge(UInt16(a), UInt16(b))
            if edgeCount[key] == nil {
                report.fail(.constraints,
                            "constraint edge (\(key.a),\(key.b)) is absent from the triangulation")
            }
        }

        // I10 — no triangle centroid inside a hole.
        for hole in holes {
            for (a, b, c) in tris {
                let centroid = (points[a] + points[b] + points[c]) / 3
                if MeshKernel.pointInRing(points: points, ring: hole, centroid) {
                    report.fail(.trianglesInHoles, "triangle (\(a),\(b),\(c)) lies inside a hole")
                    break
                }
            }
        }

        // I11 — the UInt16 index ceiling the file format imposes.
        if n > Int(UInt16.max) {
            report.fail(.indexCeiling,
                        "\(n) vertices exceeds the UInt16 index limit of \(UInt16.max)")
        }

        return report
    }

    /// Fraction of the region actually covered. 1.0 is watertight.
    static func coverageRatio(
        points: [SIMD2<Float>], triangles: [UInt16], outer: [Int], holes: [[Int]] = []
    ) -> Double {
        var covered = 0.0
        for t in stride(from: 0, to: triangles.count, by: 3) {
            let a = Int(triangles[t]), b = Int(triangles[t + 1]), c = Int(triangles[t + 2])
            guard points.indices.contains(a), points.indices.contains(b),
                  points.indices.contains(c) else { continue }
            covered += abs(MeshPredicates.signedArea2(points[a], points[b], points[c])) / 2.0
        }
        var target = MeshKernel.polygonArea(points: points, ring: outer)
        for hole in holes { target -= MeshKernel.polygonArea(points: points, ring: hole) }
        return target > 0 ? covered / target : 0.0
    }

    private static func edgeIsBefore(_ lhs: MeshEdge, _ rhs: MeshEdge) -> Bool {
        lhs.a != rhs.a ? lhs.a < rhs.a : lhs.b < rhs.b
    }
}
